### PR TITLE
 Add, use and test `VerifyWeakSubjectivityRoot`

### PR DESCRIPTION
**Pull request by terencechain**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

This PR adds in the weak subjectivity check as outlined in:
https://github.com/ethereum/eth2.0-specs/blob/master/specs/phase0/weak-subjectivity.md#weak-subjectivity-sync-procedure

This weak subjectivity check is currently opt-out. It'll switch to opt-in with the closure of issue #7342 

**Other notes for review**
Tested run time under various conditions

1.) ws block not in db:
Screen Shot 2020-09-24 at 2 21 34 PM

2.) ws block in db but incorrect epoch:
Screen Shot 2020-09-24 at 2 22 03 PM

3.) ws check working during initial syncing
Screen Shot 2020-09-24 at 2 22 48 PM

4.) ws check working for an almost caught up node
Screen Shot 2020-09-24 at 2 24 02 PM

### CODE REVIEW

File: beacon-chain/blockchain/service.go [hunk 1]
```diff
@@ -184,6 +191,12 @@ func (s *Service) Start() {
 		s.prevFinalizedCheckpt = stateTrie.CopyCheckpoint(finalizedCheckpoint)
 		s.resumeForkChoice(justifiedCheckpoint, finalizedCheckpoint)
 
+		if err := s.VerifyWeakSubjectivityRoot(s.ctx); err != nil {
+			// Exit run time if the node failed to verify weak subjectivity checkpoint.
+			log.Fatalf("Could not verify weak subjectivity checkpoint: %v", err)
+			return
```
Comment (rauljordan) [inline beacon-chain/blockchain/service.go:197]: Superflous
```suggestion
```

File: beacon-chain/blockchain/weak_subjectivity_checks.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	"github.com/prysmaticlabs/prysm/beacon-chain/db/filters"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// VerifyWeakSubjectivityRoot verifies the weak subjectivity root in the service struct.
+// Reference design: https://github.com/ethereum/eth2.0-specs/blob/master/specs/phase0/weak-subjectivity.md#weak-subjectivity-sync-procedure
+func (s *Service) VerifyWeakSubjectivityRoot(ctx context.Context) error {
+	// TODO(7342): Remove the following to fully use weak subjectivity in production.
+	if len(s.wsRoot) == 0 && s.wsEpoch == 0 {
```
Comment (rauljordan) [inline beacon-chain/blockchain/weak_subjectivity_checks.go:17]: should this be an || ?

File: beacon-chain/blockchain/weak_subjectivity_checks.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+	}
+	for _, root := range roots {
+		if r == root {
+			log.Info("Weak subjectivity check has passed")
+			s.wsVerified = true
+			return nil
+		}
+	}
+
+	return fmt.Errorf("node does not have root in db corresponds to epoch: %#x %d", r, s.wsEpoch)
```
Comment (rauljordan) [inline beacon-chain/blockchain/weak_subjectivity_checks.go:59]:
```suggestion
	return fmt.Errorf("node does not have root in db corresponding to epoch: %#x %d", r, s.wsEpoch)
```

File: beacon-chain/blockchain/weak_subjectivity_checks.go (hunk 2 above)
Comment (terencechain) [reply beacon-chain/blockchain/weak_subjectivity_checks.go]: Either way is fine. This is not going to be production code.  I changed it to `| |`

**Review by farazdagi — APPROVED** [review]
looks great!

**Pull request merged**